### PR TITLE
Allow users to provide a client web socket instance for AMQP DPS device client

### PR DESCRIPTION
**Pull request by timtay-microsoft**
This allows users to set the remote certificate validation callback as long as they target a .NET version that supports that feature.

Also add the web socket idle timeout option back into the AMQP options bundle

### CODE REVIEW

File: provisioning/device/src/TransportSettings/ProvisioningClientAmqpSettings.cs [hunk 1]
```diff
@@ -36,6 +36,24 @@ public ProvisioningClientAmqpSettings(ProvisioningClientTransportProtocol transp
...
+        /// The client web socket to use when communicating over web sockets.
+        /// </summary>
+        /// <remarks>
+        /// This option is ignored for TCP connections.
+        /// <para>
+        /// If not provided, a client web socket instance will be created for you based on the other
+        /// settings provided in this class.
+        /// </para>
+        /// </remarks>
+        public ClientWebSocket ClientWebSocket { get; set; } = new ClientWebSocket();
```
Comment (drwill-ms) [inline provisioning/device/src/TransportSettings/ProvisioningClientAmqpSettings.cs:49]: Any migration doc updates?

File: provisioning/device/src/TransportSettings/ProvisioningClientAmqpSettings.cs [hunk 2]
```diff
@@ -36,6 +36,24 @@ public ProvisioningClientAmqpSettings(ProvisioningClientTransportProtocol transp
...
+
+        /// <summary>
+        /// A keep-alive for the transport layer in sending ping/pong control frames when using web sockets.
+        /// </summary>
+        /// <seealso href="https://docs.microsoft.com/dotnet/api/system.net.websockets.clientwebsocketoptions.keepaliveinterval"/>
+        public TimeSpan? WebSocketKeepAlive { get; set; }
+
         internal override ProvisioningClientTransportSettings Clone()
```
Comment (drwill-ms) [inline provisioning/device/src/TransportSettings/ProvisioningClientAmqpSettings.cs:57]: Add to clone method?

File: provisioning/device/src/TransportSettings/ProvisioningClientTransportSettings.cs [hunk 3]
```diff
@@ -67,7 +68,9 @@ public abstract class ProvisioningClientTransportSettings
         /// If incorrectly implemented, your device may fail to connect to IoT hub and/or be open to security vulnerabilities.
         /// <para>
         /// This feature is only applicable for HTTP, MQTT over TCP, MQTT over web socket, AMQP
-        /// over TCP. AMQP web socket communication does not support this feature.
+        /// over TCP. AMQP web socket communication does not support this feature. For users who want
+        /// this support over AMQP websocket, you must instead provide a <see cref="ClientWebSocket"/>
+        /// instance with the desired callback set.
         /// </para>
         /// </remarks>
         public RemoteCertificateValidationCallback RemoteCertificateValidationCallback { get; set; } = DefaultRemoteCertificateValidation;
```
Comment (drwill-ms) [inline provisioning/device/src/TransportSettings/ProvisioningClientTransportSettings.cs:76]: What happens if this is set to null?

File: provisioning/device/src/TransportSettings/ProvisioningClientAmqpSettings.cs (hunk 1 above)
Comment (timtay-microsoft) [reply provisioning/device/src/TransportSettings/ProvisioningClientAmqpSettings.cs:49]: Do we add new features to the migration doc? These didn't exist in v1

File: provisioning/device/src/TransportSettings/ProvisioningClientTransportSettings.cs (hunk 3 above)
Comment (timtay-microsoft) [reply provisioning/device/src/TransportSettings/ProvisioningClientTransportSettings.cs:76]: Added the required null checks in the HTTP/MQTT/AMQP layers where they were missing

File: provisioning/device/src/TransportSettings/ProvisioningClientAmqpSettings.cs (hunk 1 above)
Comment (drwill-ms) [reply provisioning/device/src/TransportSettings/ProvisioningClientAmqpSettings.cs:49]: We call them out, yes.

**Review by drwill-ms — APPROVED** [review]

**Comment by tmahmood-microsoft** [comment]
Please keep these changes. I will remove my duplicate implementation

File: provisioning/device/src/TransportSettings/ProvisioningClientTransportSettings.cs [hunk 4]
```diff
@@ -67,7 +68,9 @@ public abstract class ProvisioningClientTransportSettings
         /// If incorrectly implemented, your device may fail to connect to IoT hub and/or be open to security vulnerabilities.
         /// <para>
         /// This feature is only applicable for HTTP, MQTT over TCP, MQTT over web socket, AMQP
-        /// over TCP. AMQP web socket communication does not support this feature.
+        /// over TCP. AMQP web socket communication does not support this feature. For users who want
+        /// this support over AMQP websocket, you must instead provide a <see cref="ClientWebSocket"/>
+        /// instance with the desired callback set.
```
Comment (abhipsaMisra) [inline provisioning/device/src/TransportSettings/ProvisioningClientTransportSettings.cs:73]:
```suggestion
        /// instance with the desired callback and other websocket options (eg. proxy, keep-alive etc.) set.
```

**Review by abhipsaMisra — APPROVED** [review]

**Review by brycewang-microsoft — APPROVED** [review]

**Pull request merged**